media_types/mod.ts

```diff
@@ -0,0 +1,407 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+/** Provides utility functions for media types.
+ *
+ * This API is inspired by the GoLang [`mime`](https://pkg.go.dev/mime) package
+ * and [jshttp/mime-types](https://github.com/jshttp/mime-types).
+ *
+ * @module
+ */
+
+import db from "https://raw.githubusercontent.com/jshttp/mime-db/v1.52.0/db.json" assert {
+  type: "json",
+};
+import { extname } from "../path/mod.ts";
+import {
+  consumeMediaParam,
+  decode2331Encoding,
+  isIterator,
+  isToken,
+  needsEncoding,
+} from "./_util.ts";
+
+interface DBEntry {
+  source: string;
+  compressible?: boolean;
+  charset?: string;
+  extensions?: string[];
+}
+
+type KeyOfDb = keyof typeof db;
+
+/** A map of extensions for a given media type. */
+export const extensions = new Map<string, string[]>();
+
+/** A map of the media type for a given extension */
+export const types = new Map<string, KeyOfDb>();
+
+/** Internal function to populate the maps based on the Mime DB. */
+(function populateMaps(): void {
+  const preference = ["nginx", "apache", undefined, "iana"];
+
+  for (const type of Object.keys(db) as KeyOfDb[]) {
+    const mime = db[type] as DBEntry;
+    const exts = mime.extensions;
+
+    if (!exts || !exts.length) {
+      continue;
+    }
+
+    extensions.set(type, exts);
+
+    for (const ext of exts) {
+      const current = types.get(ext);
+      if (current) {
+        const from = preference.indexOf((db[current] as DBEntry).source);
+        const to = preference.indexOf(mime.source);
+
+        if (
+          current !== "application/octet-stream" &&
+          (from > to ||
+            // @ts-ignore work around denoland/dnt#148
+            (from === to && current.startsWith("application/")))
+        ) {
+          continue;
+        }
+      }
+
+      types.set(ext, type);
+    }
+  }
+})();
+
+/** Given an extension or media type, return a full `Content-Type` or
+ * `Content-Disposition` header value.
```

Comment:
Accepting both extension and media type feels strange and confusing to me. What is the benefit of accepting both? How about splitting the API or dropping the media type as input? (Currently what we need in file_server implementation is only `extension -> content type` function)

Reply:
A lot of code based on `media_type` is written to accept both.  For example with oak, you can set the response body type to `html`/`json`/`text/html`/`application/json` to have the `Content-Type` set correctly.  This gives a path for less breaking changes as well as a migration path for people already consuming an API.

Having it clearly documented I think avoids any confusion.

Reply:
Ok. Makes sense to me.